why_postgres.mdx

```diff
@@ -0,0 +1,124 @@
+
+If file attachments or full-text search or vector/hybrid search are important for your application, Xata will likely offer you a better platform to build on top than a classical DBaaS. And it's only the beginning, we are planning to have similar tightly-integrated solutions for time-series, intelligent caching, queuing, and more.
+
+## The implementation
+
+One key concept in the Xata implementation is the notion of _logical database branches_. A database can have multiple branches, and always has at least one, called by default `main`. A database branch is represented in Postgres as a [schema](https://www.postgresql.org/docs/current/ddl-schemas.html), and can be allocated on one of many Postgres clusters around the world, as well as moved from one cluster to another.
+
+This results in a level of decoupling between the logical databases and branches as experienced by the application, and the underlying Postgres instances where they are running. Because the clusters are already provisioned, logical databases and branches can be started instantly and allocated and moved depending on the business needs. This unlocks some interesting use-cases, like sharding per tenant where every tenant is a branch, or using a logical database for each micro-service, or automatically creating developer branches for each pull request. Or, and this is important for our free tier, have many small databases and branches share the same relatively large underlying Postgres cluster.
+
+![Database branches scheduled on various Postgres clusters](images/branches-clusters.png)
+
+In order to make this possible, however, we had to solve several challenges: route traffic to the correct cluster based on metadata about the database branch, ensure that a branch level connection can only access that single branch, ensure that attackers can't call dangerous SQL functions, and limit noisy-neighbour issues in the shared clusters.
+
+#### SQL Proxy
+
+The key to the solution for most of these issues is our SQL Proxy technology. While there are many Postgres proxies available (e.g. pgbouncer), the Xata SQL Proxy not only speaks the Postgres wire protocol, but also understands the SQL being executed at a fairly deep level. It can, for example, know which tables are accessed, which columns are requests, which functions are being called, with which parameters, and so on. This allows us to block actions that we know would be dangerous for the Postgres cluster. While the security is mostly ensured via Postgres roles, the Proxy can enforce rate and concurrency limits on shared clusters (we don't do that for dedicated clusters).
```

Comment:
```suggestion
The key to the solution for most of these issues is our SQL Proxy technology. While there are many Postgres proxies available (e.g. pgbouncer), the Xata SQL Proxy not only speaks the Postgres wire protocol, but also understands the SQL being executed at a fairly deep level. It can, for example, know which tables are accessed, which columns are requested, which functions are being called, with which parameters, and so on. This allows us to block actions that we know would be dangerous for the Postgres cluster. While the security is mostly ensured via Postgres roles, the Proxy can enforce rate and concurrency limits on shared clusters (we don't do that for dedicated clusters).
```